nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/parameter/StandardParameterContext.java

```diff
@@ -0,0 +1,279 @@
+import org.apache.nifi.authorization.resource.Authorizable;
+import org.apache.nifi.authorization.resource.ResourceFactory;
+import org.apache.nifi.authorization.resource.ResourceType;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.controller.ComponentNode;
+import org.apache.nifi.controller.ProcessorNode;
+import org.apache.nifi.controller.PropertyConfiguration;
+import org.apache.nifi.controller.service.ControllerServiceNode;
+import org.apache.nifi.controller.service.ControllerServiceState;
+import org.apache.nifi.groups.ProcessGroup;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class StandardParameterContext implements ParameterContext {
+    private static final Logger logger = LoggerFactory.getLogger(StandardParameterContext.class);
+
+    private final String id;
+    private final ParameterReferenceManager parameterReferenceManager;
+
+    private String name;
+    private long version = 0L;
+    private final Map<ParameterDescriptor, Parameter> parameters = new LinkedHashMap<>();
+    private volatile String description;
+
+    private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
+    private final Lock readLock = rwLock.readLock();
+    private final Lock writeLock = rwLock.writeLock();
+
+
+    public StandardParameterContext(final String id, final String name, final ParameterReferenceManager parameterReferenceManager) {
+        this.id = Objects.requireNonNull(id);
+        this.name = Objects.requireNonNull(name);
+        this.parameterReferenceManager = parameterReferenceManager;
+    }
+
+    @Override
+    public String getIdentifier() {
+        return id;
+    }
+
+    public String getName() {
+        readLock.lock();
+        try {
+            return name;
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    public void setName(final String name) {
+        writeLock.lock();
+        try {
+            this.version++;
+            this.name = name;
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void setDescription(String description) {
+        this.description = description;
+    }
+
+    @Override
+    public String getDescription() {
+        return description;
+    }
+
+    public void setParameters(final Set<Parameter> updatedParameters) {
+        writeLock.lock();
+        try {
+            this.version++;
+
+            verifyCanSetParameters(updatedParameters);
+
+            for (final Parameter parameter : updatedParameters) {
+                if (parameter.getValue() == null) {
+                    parameters.remove(parameter.getDescriptor());
```

Comment:
When a user opts to update a sensitive parameter, the front end does not have the value to re-submit. As a result, the request would be mis-interrupted as a request to delete instead of a request to update the description. What are your thoughts on only considering the incoming request as a delete request when the description is not specified? The name and the sensitive flag are technically immutable so these could be considered optional. If the description is specified then update the parameter. If the parameter is sensitive, then only update the value if the value is non-null in the incoming request. I believe this convention is something that we could easily document in the annotations in the DTO and would not be ambiguous.

Reply:
That's a good call. I had thought about the lack of the sensitive property value on the client side but hadn't come up with a smart way of handling it. I like this approach.

Reply:
Alternatively, as we expect at least the sensitive flag to always be populated, do we want to say that the front end should pass a fully specified object for parameter creation or update, and that just passing the parameter name with no other field indicates deletion?

That is...

**Indicate update value of param1:**
```
{
    "name": "param1"
    "sensitive": "true"
    "description": "My existing, unchanged description."
    "value": "newSecretValue"
}
```

**Indicate update description of param1 (don't modify value):**
```
{
    "name": "param1"
    "sensitive": "true"
    "description": "My updated description"
}
```

**Indicate delete param1:**
```
{
    "name": "param1"
}
```

And sensitivity and name are both immutable, right?

Reply:
Thanks for commenting @kevdoran. You are correct and I think what you're proposing here will work well.